### PR TITLE
Improved REPL printing for GroupedDataFrames

### PR DESCRIPTION
**Pull request by Jollywatt**
This is a small enhancement to make `GroupedDataFrame`s display so that they fit in one screen, like regular `DataFrame`s with many rows.

Before, the first and last groups of a `GroupedDataFrame` would be displayed at full size; now, the two groups are made smaller (using the `:displaysize` parameter in `IOContext`) so that together they fill the required line height.

If the first group is smaller than half the available REPL height, and the last group is larger, or vice versa, then the smaller group is shown in full and the larger one is squashed (displayed with missing rows). This means that the group with the most rows takes up the most room visually, even after some of its rows are skipped.

See the `"GroupedDataFrame displaysize test"` test set for example output.

### CODE REVIEW

**Comment by bkamins** [comment]
Makes sense. I will wait for @ronisbr to review first as he is maintaining all display stuff 😄.

**Comment by bkamins** [comment]
@ronisbr - could you please review this PR (especially in terms of how it interacts with PrettyTables.jl kwargs) while we are waiting for HTML PR to be merged. Then I will review it. Thank you!

**Comment by ronisbr** [comment]
Sorry! I completely missed this one.

**Review by ronisbr — COMMENTED** [review]
It looks good to me! I would just add tests for corner cases in `displaysize`. For example, if we are printing to a file, `dislaysize` will take the values from `LINES` and `COLUMNS`:

```julia
julia> io = IOBuffer()
IOBuffer(data=UInt8[...], readable=true, writable=true, seekable=true, append=false, size=0, maxsize=Inf, ptr=1, mark=-1)

julia> ENV["LINES"] = -1
-1

julia> displaysize(io)
(-1, 80)
```

In this case, I suppose the data frames will be printed without limits given how `show` works. However, it will important to have tests for those cases with invalid options.

Another corner case that I always have problems when dealing with limited display is when the display is just too small. What will happen if the display here has only 3 lines? Will the algorithm break at some point? I think `h = div(v, 2)` can return a number equal or lower than 0, which will print the entire DataFrame. Can you check this please?

**Comment by Jollywatt** [comment]
Hi @ronisbr, thanks for the comment!
I made sure the printing obeys “print squashed if `h` is small but positive, but print everything if `h` is invalid (zero or nagative)” and added tests. (This seems to be how ungrouped data frames behave.) Previously, `h = 1` and `h = 2` were edge cases where one group would be printed in full bu not the other — good catch!

**Comment by bkamins** [comment]
@Jollywatt  - unfortunately due to Tables.jl 1.8 release we needed to make some changes in the internals of DataFrames.jl. Can you please rebase (or merge if rebasing is problematic) this PR against `main` branch as otherwise CI will fail for this PR.

File: src/groupeddataframe/show.jl
Comment (bkamins) [inline src/groupeddataframe/show.jl:41]: @ronisbr - do you use the same pattern in PrettyTables.jl? (to ensure consistency)

@Jollywatt - should not we use the `get(io, :limit, false)` test first and only implement the logic you propose if the `:limit` property is set? And add tests for both cases.

Note the following for standard printing of data frames:
```
julia> df = DataFrame(rand(20, 5), :auto) # display sets limit
20×5 DataFrame
 Row │ x1        x2         x3          x4         x5
     │ Float64   Float64    Float64     Float64    Float64
─────┼───────────────────────────────────────────────────────
   1 │ 0.799571  0.711126   0.0829037   0.153365   0.65724
   2 │ 0.013505  0.516158   0.289836    0.772033   0.236464
   3 │ 0.539347  0.866731   0.527391    0.763274   0.445696
  ⋮  │    ⋮          ⋮          ⋮           ⋮          ⋮
  18 │ 0.285917  0.83837    0.412981    0.283734   0.734432
  19 │ 0.539428  0.281714   0.503087    0.236329   0.980165
  20 │ 0.19463   0.818894   0.080396    0.274849   0.678159
                                              14 rows omitted

julia> show(stdout, df) # stdout does not have limit
20×5 DataFrame
 Row │ x1        x2         x3          x4         x5
     │ Float64   Float64    Float64     Float64    Float64
─────┼────────────────────────────────────────────────────────
   1 │ 0.799571  0.711126   0.0829037   0.153365   0.65724
   2 │ 0.013505  0.516158   0.289836    0.772033   0.236464
   3 │ 0.539347  0.866731   0.527391    0.763274   0.445696
   4 │ 0.220234  0.94367    0.740741    0.141462   0.121854
   5 │ 0.32305   0.267344   0.519681    0.331716   0.396574
   6 │ 0.127974  0.282572   0.640234    0.841006   0.526737
   7 │ 0.322138  0.925825   0.00683361  0.0276174  0.140838
   8 │ 0.832967  0.744335   0.808838    0.516439   0.219536
   9 │ 0.895046  0.853179   0.487567    0.81502    0.721893
  10 │ 0.236999  0.481331   0.310607    0.0200533  0.00837921
  11 │ 0.142544  0.310369   0.121286    0.781769   0.0948759
  12 │ 0.923275  0.205102   0.097269    0.806569   0.273017
  13 │ 0.177222  0.266212   0.965611    0.345097   0.745195
  14 │ 0.157075  0.851299   0.121473    0.281857   0.1338
  15 │ 0.938703  0.183045   0.769088    0.562905   0.0955595
  16 │ 0.592258  0.387368   0.912301    0.292013   0.401606
  17 │ 0.949321  0.0872963  0.421829    0.0586528  0.39613
  18 │ 0.285917  0.83837    0.412981    0.283734   0.734432
  19 │ 0.539428  0.281714   0.503087    0.236329   0.980165
  20 │ 0.19463   0.818894   0.080396    0.274849   0.678159
julia> show(IOContext(stdout, :limit => true), df) # but you can manually enable limit
20×5 DataFrame
 Row │ x1        x2         x3          x4         x5
     │ Float64   Float64    Float64     Float64    Float64
─────┼───────────────────────────────────────────────────────
   1 │ 0.799571  0.711126   0.0829037   0.153365   0.65724
   2 │ 0.013505  0.516158   0.289836    0.772033   0.236464
   3 │ 0.539347  0.866731   0.527391    0.763274   0.445696
  ⋮  │    ⋮          ⋮          ⋮           ⋮          ⋮
  18 │ 0.285917  0.83837    0.412981    0.283734   0.734432
  19 │ 0.539428  0.281714   0.503087    0.236329   0.980165
  20 │ 0.19463   0.818894   0.080396    0.274849   0.678159
                                              14 rows omitted
```

**Comment by Jollywatt** [comment]
@bkamins I hope I did that right — I synced my fork using the Github UI, which made a merge, not a rebase.

**Comment by bkamins** [comment]
merge should be OK.

File: test/show.jl
Comment (bkamins) [inline test/show.jl:167]: can you please add tests for `GroupedDataFrame` with 0 and with 1 group?

Comment (Jollywatt) [reply test/show.jl:167]: @bkamins  Ok, will do. Quick side question: the commit with which I merged (`d6277e40d`, message `Synchronize ByRow and Tables.subset with Tables.jl 1.8 (#3158)`) doesn’t precompile (throws `ByRow not defined`). Not sure how to “un-sync” my fork — should I wait for you to commit a fix and then resync my fork?

Comment (bkamins) [reply test/show.jl:167]: Have you made sure to update Tables.jl to 1.8.1?
What is the problem exactly on your end?

All works OK on my machine:
```
(@v1.8) pkg> activate .
  Activating new project at `~/WORK`

(WORK) pkg> add DataFrames#main
    Updating git-repo `https://github.com/JuliaData/DataFrames.jl.git`
    Updating registry at `~/.julia/registries/General`
    Updating git-repo `https://github.com/JuliaRegistries/General.git`
   Resolving package versions...
    Updating `~/WORK/Project.toml`
  [a93c6f00] + DataFrames v1.4.0 `https://github.com/JuliaData/DataFrames.jl.git#main`
    Updating `~/WORK/Manifest.toml`
  [34da2185] + Compat v4.2.0
  [a8cc5b0e] + Crayons v4.1.1
  [9a962f9c] + DataAPI v1.10.0
  [a93c6f00] + DataFrames v1.4.0 `https://github.com/JuliaData/DataFrames.jl.git#main`
  [864edb3b] + DataStructures v0.18.13
  [e2d170a0] + DataValueInterfaces v1.0.0
  [59287772] + Formatting v0.4.2
  [41ab1584] + InvertedIndices v1.1.0
  [82899510] + IteratorInterfaceExtensions v1.0.0
  [e1d29d7a] + Missings v1.0.2
  [bac558e1] + OrderedCollections v1.4.1
  [2dfb63ee] + PooledArrays v1.4.2
⌃ [08abe8d2] + PrettyTables v1.3.1
  [189a3867] + Reexport v1.2.2
  [a2af1166] + SortingAlgorithms v1.0.1
  [3783bdb8] + TableTraits v1.0.1
  [bd369af6] + Tables v1.8.1
  [56f22d72] + Artifacts
  [2a0f44e3] + Base64
  [ade2ca70] + Dates
  [9fa8497b] + Future
  [b77e0a4c] + InteractiveUtils
  [8f399da3] + Libdl
  [37e2e46d] + LinearAlgebra
  [56ddb016] + Logging
  [d6f4376e] + Markdown
  [de0858da] + Printf
  [3fa0cd96] + REPL
  [9a3f8284] + Random
  [ea8e919c] + SHA v0.7.0
  [9e88b42a] + Serialization
  [6462fe0b] + Sockets
  [2f01184e] + SparseArrays
  [10745b16] + Statistics
  [8dfed614] + Test
  [cf7118a7] + UUIDs
  [4ec0a83e] + Unicode
  [e66e0078] + CompilerSupportLibraries_jll v0.5.2+0
  [4536629a] + OpenBLAS_jll v0.3.20+0
  [8e850b90] + libblastrampoline_jll v5.1.1+0
        Info Packages marked with ⌃ have new versions available
Precompiling project...
  1 dependency successfully precompiled in 18 seconds. 19 already precompiled.

julia> using DataFrames

julia>
```

Comment (Jollywatt) [reply test/show.jl:167]: That might have been it. I re-cloned my fork and re-`Pkd.dev`ed it — working now!

File: src/groupeddataframe/show.jl
Comment (bkamins) [inline src/groupeddataframe/show.jl:47]: @ronisbr - there is also an additional kwarg in PrettyTables.jl that governs number of rows to be printed in a table. Can you please remind it its name? We should check how passing it interacts with this PR.

Comment (bkamins) [reply src/groupeddataframe/show.jl:47]: One for sure I now see is `display_size`.

**Comment by bkamins** [comment]
In summary - the PR mostly looks good (it works OK under standard settings). However, please carefully consider `:limit` setting of `IOContext` and possible kwargs that can be passed to PrettyTables.jl so make sure we do not hit some problem in such cases.

File: src/groupeddataframe/show.jl
Comment (Jollywatt) [reply src/groupeddataframe/show.jl:47]: I should point out that change to line 47 doesn't change any behaviour: arguments like `allrows` get forwarded to `show` methods later on. (All `!allrows && …` does is save a few arithmetic operations in the case where we know that we don't need to calculate heights.) So if there are other `kwargs` that override `allrows`, then it might be better to always set the default `display_size` like before.

Comment (bkamins) [reply src/groupeddataframe/show.jl:47]: Before I did not suggest to make a change to check `allrows`, but rather to check `get(io, :limit, false)`. However, looking at the code - maybe it is OK, as if `:limit` is `false` then display size will be ignored in `show` for groups. Just please confirm that we have a test for this.

Comment (Jollywatt) [reply src/groupeddataframe/show.jl:47]: Oh, does `:limit` take precedence over `allrows`? I have added a single test for whether `allrows=true` works even when display size is small, but don't have tests where I set `:limit` directly.

Comment (bkamins) [reply src/groupeddataframe/show.jl:47]: > does `:limit` take precedence over `allrows`?

It depends how you define *precedence*. `allrows` will override `:limit` if `allrows` is set to `true`.
However, you can have `allrows=false` and `:limit=false`. We should make sure that even if we pass custom `display_size` to `show` internally all works OK (I think it will, but let us make sure).

**Comment by bkamins** [comment]
Also documentation needs to be updated to reflect the changes rules of printing of `GroupedDataFrame`.

File: src/groupeddataframe/show.jl
Comment (ronisbr) [reply src/groupeddataframe/show.jl:41]: > @ronisbr - do you use the same pattern in PrettyTables.jl? (to ensure consistency) 

Yes, kind of. The idea is the same, but I check for `:limit` if the user did not pass a specific keyword to crop the output table.

File: src/groupeddataframe/show.jl
Comment (ronisbr) [reply src/groupeddataframe/show.jl:47]: > @ronisbr - there is also an additional kwarg in PrettyTables.jl that governs number of rows to be printed in a table. Can you please remind it its name? We should check how passing it interacts with this PR.

You can select the number of rows that will be printed if `crop` is set to `:both` or `:vertical`. In this case, the number of printed rows (including header, table dividers, etc) are selected based on the display size.

**Comment by ronisbr** [comment]
From my side, I do not see problems related to kwargs in PrettyTables. The arguments `allrows` and `allcols` are converted correctly to the one in PrettyTables that limit the number of rows and columns to be printed. Of course, the user can break it badly by passing strange arguments to `show`, but this is the side effect of the power to customize the output.

**Comment by bkamins** [comment]
OK - so we can drop `!allrows` check then. @Jollywatt - can you then please remove it + update documentation examples so that documenter passes correctly?

**Comment by bkamins** [comment]
@nalimilan - do you have time to have a look at this PR (and as usual spot the tiny details that I usually miss; thank you!)

File: src/groupeddataframe/show.jl
Comment (nalimilan) [inline src/groupeddataframe/show.jl:79]:
```suggestion
             truncate=truncate, display_size=(h1, w), kwargs...)
```

File: src/groupeddataframe/show.jl
Comment (nalimilan) [inline src/groupeddataframe/show.jl:97]:
```suggestion
             truncate=truncate, display_size=(h2, w), kwargs...)
```

File: test/show.jl
Comment (nalimilan) [inline test/show.jl:347]: Probably not a big deal, but in this case we could avoid printing this part as it doesn't add any information:
```
  ⋮  │   ⋮      ⋮      ⋮
                         5 rows omitted
```

Comment (ronisbr) [reply test/show.jl:347]: If you want to omit this information, you need to pass `show_omitted_cell_summary = false` to PrettyTables.jl.

Comment (bkamins) [reply test/show.jl:347]: I would leave it as is for now (but I do not have a strong preference). The point is that the message potentially also contains information about how many columns are not printed.

**Comment by bkamins** [comment]
@Jollywatt - do you have time to have a look at this PR and finalize it. Thank you!

Soon we will make a feature freeze for DataFrames.jl 1.4 release (after this - the PR would still be merged but would be released later).

**Comment by Jollywatt** [comment]
@bkamins Ok, done (I think)! Let me know if there’s anything else to do:)

File: src/groupeddataframe/show.jl
Comment (bkamins) [reply src/groupeddataframe/show.jl:41]: I think it does not have to, see e.g. https://github.com/JuliaLang/julia/blob/d5cde865f24d2c3e7041aee8ea464eb6b6045a2a/base/stream.jl#L564 (and custom targets could define similar rules).

OTOH, I do not see a situation when what is proposed now would be incorrect.

**Review by bkamins — APPROVED** [review]

**Review by bkamins — CHANGES_REQUESTED** [review]
tests are failing

File: test/show.jl
Comment (bkamins) [inline test/show.jl:387]:
```suggestion
```

File: src/groupeddataframe/show.jl
Comment (bkamins) [inline src/groupeddataframe/show.jl:65]:
```suggestion
```

File: src/groupeddataframe/show.jl
Comment (bkamins) [inline src/groupeddataframe/show.jl:83]:
```suggestion
```

File: src/groupeddataframe/show.jl
Comment (bkamins) [inline src/groupeddataframe/show.jl:80]:
```suggestion
```

File: src/groupeddataframe/show.jl
Comment (nalimilan) [reply src/groupeddataframe/show.jl:41]: Neither DataFrames nor PrettyTables use that pattern AFAICT. Julia Base doesn't. So better be consistent.

Comment (bkamins) [reply src/groupeddataframe/show.jl:41]: DataFrames.jl uses this pattern in https://github.com/JuliaData/DataFrames.jl/blob/8acb679cafc44c2d6b524fbaf4df7809f9ded8f2/src/abstractdataframe/io.jl#L389

Blame shows that I added it 3 years ago in https://github.com/JuliaData/DataFrames.jl/pull/1761. The reason why it was added is https://github.com/JuliaData/DataFrames.jl/pull/1761#discussion_r275116264 (you had the same comment there as here). @nalimilan - so what do we do?

**Comment by Jollywatt** [comment]
> tests are failing

I’ve been trying to debug those failures but haven’t gotten to the bottom of them. It seems like `PrettyTables` got updated?

**Comment by bkamins** [comment]
docs build is failing.

**Comment by bkamins** [comment]
Regarding failing CI:

@ronisbr - this is a bug in PrettyTables.jl. You use `@inbounds` in several places in the code, so this does not get detected in normal usage, but tests catch it (as they force bounds checking).

The offending operation is:
```
num_lines_in_row[1:num_header_rows]
```
where in the test `num_header_rows` is 2, while `num_lines_in_row` is 1.

This is in `_count_header_lines` function.

The issue is most likely some non standard table printing heights that get passed in this PR to PrettyTables.jl.

So the fix should have two parts:
* personally - I would remove `@inbounds` here as it seems your code does not ensure that `@inbounds` is safe (so it should not be used; also - most likely it does not save much time - but you will know better); (in general please consider in whole PrettyTables.jl if you indeed need to use `@inbounds` in places where you use them)
* probably some more checking of passed arguments needs to be done in PrettyTables.jl and they should be corrected if they are e.g. too restrictive

**Comment by ronisbr** [comment]
Oops, sorry about that! I always assumed that the header is always rendered. However, given the huge rewriting of PrettyTables in v2, I forgot to add some checks. In this case, the number of rendered lines was set to the number of display lines (1), which is lower than the number of header lines (2). This must never happen. I fixed this bug in `master` and will release a new minor version as soon as the test finishes.

**Comment by ronisbr** [comment]
@Jollywatt by the way, the output will change. Notice that the header will always be printed, no matter how small the display is.

**Comment by bkamins** [comment]
I think it is OK that the header is always printed.

**Comment by bkamins** [comment]
@Jollywatt - after PrettyTables.jl changes docstrings + testset outputs need to be updated. Thank you!
After this I think all is good to merge.

**Comment by Jollywatt** [comment]
@bkamins I’ve fixed the display logic and tests to work for PrettyTables 2.1.1, and fixed the doctests. Hopefully I haven’t missed anything:) (Apologies for the highly nonlinear git history…!)

**Comment by bkamins** [comment]
Thank you! I need to review the PR before merging from scratch anyway so it is not a problem that history is complex.

**Review by bkamins — APPROVED** [review]

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**
